File: tests/test_write_diff_integration.c
/**
 * test_write_diff_integration.c - Integration test for Write tool with diff colorization
 *
 * Tests the complete Write tool workflow including colorized diffs when overwriting existing files
 */

#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <cjson/cJSON.h>

// Include internal header to get ConversationState definition
#include "../src/claude_internal.h"

// Define TEST_BUILD to expose internal functions
#define TEST_BUILD 1

// External functions from claude.c (exposed via TEST_BUILD)
extern int write_file(const char *path, const char *content);
extern char* read_file(const char *path);
extern int show_diff(const char *file_path, const char *original_content);
extern cJSON* tool_write(cJSON *params, ConversationState *state);



// Test helpers
static void test_write_with_diff(const char *test_name,
                                 const char *file_path,
                                 const char *original_content,
                                 const char *new_content) {
    printf("\n╔════════════════════════════════════════╗\n");
    printf("║ Test: %-32s ║\n", test_name);
    printf("╚════════════════════════════════════════╝\n\n");

    // Write original content if provided
    if (original_content) {
        if (write_file(file_path, original_content) != 0) {
            printf("❌ Failed to write original file\n");
            return;
        }

        printf("Original content:\n");
        printf("─────────────────────────────────────────\n");
        printf("%s", original_content);
        printf("─────────────────────────────────────────\n\n");
    } else {
        printf("Creating new file (no original content)\n\n");
    }

    // Perform write
    printf("Writing new content:\n");
    printf("─────────────────────────────────────────\n");
    printf("%s", new_content);
    printf("─────────────────────────────────────────\n\n");

    // Create JSON parameters for the Write tool
    cJSON *params = cJSON_CreateObject();
    cJSON_AddStringToObject(params, "file_path", file_path);
    cJSON_AddStringToObject(params, "content", new_content);

    // Create conversation state
    ConversationState state = {0};
    if (conversation_state_init(&state) != 0) {
        fprintf(stderr, "Failed to initialize conversation state\n");
        exit(1);
    }
    state.working_dir = strdup("/tmp");

    // Call the Write tool
    cJSON *result = tool_write(params, &state);

    if (result) {
        if (cJSON_HasObjectItem(result, "error")) {
            printf("❌ Write failed: %s\n", cJSON_GetObjectItem(result, "error")->valuestring);
        } else {
            printf("✓ Write completed successfully\n");
        }
        cJSON_Delete(result);
    } else {
        printf("❌ Write failed with NULL result\n");
    }

    cJSON_Delete(params);
    free(state.working_dir);
    conversation_state_destroy(&state);
    printf("\n");

    // Clean up
    unlink(file_path);
}

int main(void) {
    printf("\n╔════════════════════════════════════════════════════════════════╗\n");
    printf("║   Write Tool + Diff Colorization Integration Tests          ║\n");
    printf("╚════════════════════════════════════════════════════════════════╝\n");

    // Test 1: Overwrite existing file with simple change
    test_write_with_diff(
        "Overwrite existing file",
        "/tmp/test_write_1.txt",
        "Hello World\nThis is a test\nGoodbye World\n",
        "Hello Universe\nThis is a modified test\nGoodbye World\n"
    );

    // Test 2: Overwrite with multi-line changes
    test_write_with_diff(
        "Multi-line overwrite",
        "/tmp/test_write_2.txt",
        "Line 1: original\nLine 2: original\nLine 3: original\n",
        "Line 1: MODIFIED\nLine 2: original\nLine 3: MODIFIED\nNew Line 4: added\n"
    );

    // Test 3: Create new file (no diff should be shown)
    test_write_with_diff(
        "Create new file",
        "/tmp/test_write_3.txt",
        NULL,  // No original content
        "This is a brand new file\nWith some content\n"
    );

    // Test 4: Complete content replacement
    test_write_with_diff(
        "Complete replacement",
        "/tmp/test_write_4.txt",
        "Old content line 1\nOld content line 2\nOld content line 3\n",
        "Completely new content\nDifferent structure\nNew format\n"
    );

    // Test 5: Complex multi-line replacement
    test_write_with_diff(
        "Complex multi-line replacement",
        "/tmp/test_write_5.txt",
        "function oldFunction() {\n    console.log('old code');\n    return false;\n}",
        "function newFunction() {\n    console.log('new improved code');\n    return true;\n}\n\n// Additional helper function\nfunction helper() {\n    return 'helper result';\n}"
    );

    printf("\n╔════════════════════════════════════════╗\n");
    printf("║   All Integration Tests Completed     ║\n");
    printf("╚════════════════════════════════════════╝\n\n");

    return 0;
}

File: zig/tests/test_write_diff_integration.zig
//! tests/test_write_diff_integration.zig — Zig port of tests/test_write_diff_integration.c
//!
//! Integration tests for the Write tool covering:
//!   - Overwriting an existing file with new content
//!   - Creating a new file (no original content)
//!   - Complete content replacement
//!   - Multi-line content written correctly
//!   - Written content is exactly what was specified (byte-for-byte)
//!   - Result JSON contains status=success
//!
//! Note: The C test also checked colorised diff output when overwriting files,
//! which is a display-side concern.  The Zig Write tool does not produce diffs;
//! we verify functional correctness instead.

const std = @import("std");
const fs_mod = @import("../tools/filesystem.zig");

const executeWrite = fs_mod.executeWrite;
const executeRead = fs_mod.executeRead;

// ---------------------------------------------------------------------------
// helpers
// ---------------------------------------------------------------------------

fn makeTempFile(tmp_dir: std.testing.TmpDir, name: []const u8, content: []const u8) !void {
    try tmp_dir.dir.writeFile(name, content);
}

fn readTempFile(alloc: std.mem.Allocator, abs_dir: []const u8, name: []const u8) ![]u8 {
    const path = try std.fs.path.join(alloc, &.{ abs_dir, name });
    defer alloc.free(path);
    const f = try std.fs.openFileAbsolute(path, .{});
    defer f.close();
    return f.readToEndAlloc(alloc, 1024 * 64);
}

fn runWrite(alloc: std.mem.Allocator, json: []const u8) !fs_mod.ToolResult {
    const parsed = try std.json.parseFromSlice(std.json.Value, alloc, json, .{});
    defer parsed.deinit();
    return executeWrite(alloc, parsed.value);
}

// ---------------------------------------------------------------------------
// tests
// ---------------------------------------------------------------------------

test "Write diff integration: overwrite existing file with simple change" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    // Write original content
    try makeTempFile(tmp, "overwrite.txt", "Hello World\nThis is a test\nGoodbye World\n");

    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(alloc, &.{ tmp_path, "overwrite.txt" });
    defer alloc.free(file_path);

    const new_content = "Hello Universe\nThis is a modified test\nGoodbye World\n";
    const json = try std.fmt.allocPrint(
        alloc,
        "{{\"file_path\":\"{s}\",\"content\":\"Hello Universe\\nThis is a modified test\\nGoodbye World\\n\"}}",
        .{file_path},
    );
    defer alloc.free(json);

    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(!result.is_error);
    try std.testing.expect(std.mem.indexOf(u8, result.content, "success") != null);

    const after = try readTempFile(alloc, tmp_path, "overwrite.txt");
    defer alloc.free(after);
    try std.testing.expectEqualStrings(new_content, after);
}

test "Write diff integration: create new file" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(alloc, &.{ tmp_path, "new_file.txt" });
    defer alloc.free(file_path);

    const content = "This is a brand new file\nWith some content\n";
    const json = try std.fmt.allocPrint(
        alloc,
        "{{\"file_path\":\"{s}\",\"content\":\"This is a brand new file\\nWith some content\\n\"}}",
        .{file_path},
    );
    defer alloc.free(json);

    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(!result.is_error);

    const after = try readTempFile(alloc, tmp_path, "new_file.txt");
    defer alloc.free(after);
    try std.testing.expectEqualStrings(content, after);
}

test "Write diff integration: multi-line overwrite replaces all content" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    try makeTempFile(tmp, "multi.txt", "Line 1: original\nLine 2: original\nLine 3: original\n");

    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(alloc, &.{ tmp_path, "multi.txt" });
    defer alloc.free(file_path);

    const new_content = "Line 1: MODIFIED\nLine 2: original\nLine 3: MODIFIED\nNew Line 4: added\n";
    const json = try std.fmt.allocPrint(
        alloc,
        "{{\"file_path\":\"{s}\",\"content\":\"Line 1: MODIFIED\\nLine 2: original\\nLine 3: MODIFIED\\nNew Line 4: added\\n\"}}",
        .{file_path},
    );
    defer alloc.free(json);

    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(!result.is_error);

    const after = try readTempFile(alloc, tmp_path, "multi.txt");
    defer alloc.free(after);
    try std.testing.expectEqualStrings(new_content, after);
}

test "Write diff integration: complete content replacement" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    try makeTempFile(tmp, "replace.txt", "Old content line 1\nOld content line 2\nOld content line 3\n");

    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(alloc, &.{ tmp_path, "replace.txt" });
    defer alloc.free(file_path);

    const new_content = "Completely new content\nDifferent structure\nNew format\n";
    const json = try std.fmt.allocPrint(
        alloc,
        "{{\"file_path\":\"{s}\",\"content\":\"Completely new content\\nDifferent structure\\nNew format\\n\"}}",
        .{file_path},
    );
    defer alloc.free(json);

    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(!result.is_error);

    const after = try readTempFile(alloc, tmp_path, "replace.txt");
    defer alloc.free(after);
    try std.testing.expectEqualStrings(new_content, after);
    // Old content must be gone
    try std.testing.expect(std.mem.indexOf(u8, after, "Old content") == null);
}

test "Write diff integration: write empty content truncates file" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    try makeTempFile(tmp, "truncate.txt", "Some existing content to be erased\n");

    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(alloc, &.{ tmp_path, "truncate.txt" });
    defer alloc.free(file_path);

    const json = try std.fmt.allocPrint(
        alloc,
        "{{\"file_path\":\"{s}\",\"content\":\"\"}}",
        .{file_path},
    );
    defer alloc.free(json);

    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(!result.is_error);

    const after = try readTempFile(alloc, tmp_path, "truncate.txt");
    defer alloc.free(after);
    try std.testing.expectEqualStrings("", after);
}

test "Write diff integration: result JSON contains file_path" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(alloc, &.{ tmp_path, "check_path.txt" });
    defer alloc.free(file_path);

    const json = try std.fmt.allocPrint(
        alloc,
        "{{\"file_path\":\"{s}\",\"content\":\"hello\\n\"}}",
        .{file_path},
    );
    defer alloc.free(json);

    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(!result.is_error);
    // Result should echo back the file path
    try std.testing.expect(std.mem.indexOf(u8, result.content, "check_path.txt") != null);
}

test "Write diff integration: missing content parameter returns error" {
    const alloc = std.testing.allocator;

    const json = "{\"file_path\":\"/tmp/klawed_test_no_content.txt\"}";
    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(result.is_error);
    try std.testing.expect(std.mem.indexOf(u8, result.content, "content") != null or
        std.mem.indexOf(u8, result.content, "Missing") != null);
}

test "Write diff integration: missing file_path parameter returns error" {
    const alloc = std.testing.allocator;

    const json = "{\"content\":\"hello\"}";
    var result = try runWrite(alloc, json);
    defer result.deinit(alloc);

    try std.testing.expect(result.is_error);
}
